### PR TITLE
Fix ordering issue in analyzer callbacks for nested actions in symbol start analyzers

### LINKED ISSUE #68484: Issue in analyzer callbacks for nested actions in symbol start analyzers
Issue found by the unit test skipped in https://github.com/dotnet/roslyn/pull/68385/commits/2e0f6b90a568b1e0033f9ed785ff22c8a8fe96af.

Core issue is as follows:
1. Create a symbol start analyzer that registers nested operation actions for code within named types
2. Register symbol end action for the outer type, but not for the nested type
3. Race condition: Symbol end action callback for the outer type gets invoked before the nested operation callback for executable code within the nested type.

Note that the above does not repro if the analyzer registers a symbol end action also for the nested type.

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #68484

Race condition is detailed in the description of #68484.

Fix is to ensure that we also mark the members of nested types as dependent symbols so that the SymbolEnd actions for the outer type are executed only after all callbacks for nested type members have been done.

**Testing:** This bug was identified by the newly added CSharpUsePrimaryConstructorDiagnosticAnalyzer. There is an existing unit test that failed intermittently due to this race condition: https://github.com/CyrusNajmabadi/roslyn/blob/f3e4a27818ea2495606f12fcabd986238b66efe4/src/Analyzers/CSharp/Tests/UsePrimaryConstructor/UsePrimaryConstructorTests.cs#LL741C27-L741C62 A workaround was added to the analyzer to register a dummy SymbolEnd action for nested types to work around this bug. This PR reverts that workaround and I verified that the test passes successfully for hundreds of iterations. I also attempted to create a standalone compiler unit test, but was unable to get this race condition to repro for the test even without the product change.

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -1620,7 +1620,13 @@ async Task onSymbolAndMembersProcessedAsync(ISymbol symbol, DiagnosticAnalyzer a
...
 
                 await processContainerOnMemberCompletedAsync(symbol.ContainingNamespace, symbol, analyzer).ConfigureAwait(false);
-                await processContainerOnMemberCompletedAsync(symbol.ContainingType, symbol, analyzer).ConfigureAwait(false);
+
+                var containingType = symbol.ContainingType;
+                while (containingType != null)
+                {
+                    await processContainerOnMemberCompletedAsync(containingType, symbol, analyzer).ConfigureAwait(false);
+                    containingType = containingType.ContainingType;
+                }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1629]:
```suggestion
                for (var type = symbol.ContainingType; type != null; type = type.ContainingType)
                    await processContainerOnMemberCompletedAsync(type, symbol, analyzer).ConfigureAwait(false);
```

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for reviews.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler Ping for reviews.

**Comment by AlekseyTs** [comment]
@mavasani Unfortunately, I am not familiar with the code being modified. Could you explain the following?
- what effect the change has
- why it is appropriate to do what the modified code does
- how does the change address the problem

**Comment by mavasani** [comment]
@AlekseyTs definitely, let me try to summarize.

SymbolStart/End pair of actions allow analyzers to register a callback at a start of a symbol declaration (most often a named type symbol), then register nested callbacks within the scope of this symbol's declaration nodes for any syntax nodes or operations within these declaration nodes to accumulate analysis state and finally register a symbol end action callback that gets called after all the callbacks have been made for everything defined within this outer symbol, so it can use the accumulated state to report diagnostics. These analyzers rely on the ordering guarantee that symbol end callback is made after everything defined inside the outer symbol has received all callbacks. To ensure this ordering, whenever non-zero SymbolEndActionsCount is registered, we compute the inner dependent symbols for the outer symbol for which symbol start callback was made. This core logic is defined in the below code: https://github.com/dotnet/roslyn/blob/06534d5dc61bc3b772a1effcac725a6830fa0556/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.AnalyzerExecutionContext.cs#L138-L221

This logic was only adding the most immediate containing member symbols as dependent symbols, but not member symbols of nested types. This is sufficient for most real world analysis cases (and hence wasn't identified until now), but the analyzer that Cyrus recently added relied on gathering analysis state information even for nodes/operations within the member symbols of nested types to report the correct diagnostic in the symbol end callback for the outer type.

This change updates this dependent symbol computation logic to even consider members of nested types as dependent symbols. Additionally, the change in `AnalyzerDriver.cs` in this PR, which is executed after a symbol has been completely analyzed, now walks up the ContainingType chain to see if this was the last dependent symbol to be analyzed for this named type, and if so the corresponding named type's symbol end action gets called.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Comment by mavasani** [comment]
> LGTM (commit 3). From offline discussion, there are tests that fail in the IDE for the IDE analyzer changed in the PR if only the IDE changes are made. However, consider adding a targeted test in the analyzer layer.

I have tried adding an analyzer unit test with https://github.com/dotnet/roslyn/compare/main...mavasani:roslyn:TestForCyrus, but am unable to get it to fail without the product change. I have however verified that the IDE unit tests for [CSharpUsePrimaryConstructorDiagnosticAnalyzer.cs](https://github.com/dotnet/roslyn/pull/68490/files#diff-6ae499eb1d4910da5a1e856c0dd3b5a2ef9196d6ee7538a5dcd7477ea2855062) fail more than half the times if the compiler layer product changes in this PR are reverted.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for second review

**Comment by mavasani** [comment]
Thank you @cston

**Pull request merged**